app/models/regional_record.rb

```diff
@@ -0,0 +1,90 @@
+  validates :continent_id, presence: true, if: -> { record_scope == 'CR' }
+
+  scope :world_records, -> { where(record_scope: :WR) }
+  scope :national_records, lambda { |country_id = nil|
+    scope = where(record_scope >= :NR)
+    country_id.present? ? scope.where(country_id: country_id) : scope
+  }
+
+  scope :continental_records, lambda { |continent_id = nil|
+    scope = where(record_scope >= :CR)
+    continent_id.present? ? scope.where(continent_id: continent_id) : scope
+  }
+
+  scope :current, lambda {
+    latest = select("event_id, record_type, MAX(record_timestamp) as latest_time")
+             .group(:event_id, :record_type)
+
+    joins("INNER JOIN (#{latest.to_sql}) latest_records
+         ON regional_records.event_id = latest_records.event_id
+         AND regional_records.record_type = latest_records.record_type
+         AND regional_records.record_timestamp = latest_records.latest_time")
+  }
+
+  belongs_to :result
+
+  def self.record_for(event_id, record_type, scope, country_id: nil, continent_id: nil, date: nil)
+    query = where(event_id: event_id, record_type: record_type, record_scope: scope...)
+    query = query.where(country_id: country_id) if country_id
+    query = query.where(continent_id: continent_id) if continent_id
+    query = query.where(record_timestamp: ..date) if date
+    query.minimum(:value)
+  end
+
+  # To make this into a single query, we will need to store duplicates of every NR, WR, CR
+  # then we can do something like
+  # records = RegionalRecord.where(event_id: event_id, record_type: record_type)
+  #                           .where(record_scope: [:WR, :CR, :NR])
+  #                           .where(record_timestamp: ..timestamp)
+  #                           .where(
+  #                             scope_filter.eq(:WR)
+  #                             .or(scope_filter.eq(:CR).and(arel_table[:continent_id].eq(continent_id)))
+  #                             .or(scope_filter.eq(:NR).and(arel_table[:country_id].eq(country_id)))
+  #                           )
+  #                           .group(:record_scope)
+  #                           .minimum(:value)
+  # Not sure if worth it right now
+  def self.is_record_at_date?(value, event_id, record_type, country_id, timestamp)
+    return [true, "WR"] if record_for(event_id, record_type, :WR, date: timestamp) >= value
+
+    country = Country.c_find(country_id)
+    return [true, CONTINENT_TO_RECORD_MARKER[country.continent_id]] if record_for(event_id, record_type, :CR, continent_id: country.continent_id, date: timestamp) >= value
+
+    return [true, "NR"] if record_for(event_id, record_type, :NR, country_id: country_id, date: timestamp) >= value
+
+    [false, nil]
+  end
+
+  CONTINENT_TO_RECORD_MARKER = {
+    '_Africa' => 'AfR',
+    '_Europe' => 'ER',
+    '_North America' => 'NAR',
+    '_South America' => 'SAR',
+    '_Asia' => 'AsR',
+    '_Oceania' => 'OcR',
+  }.freeze
```

Comment:
We're storing this in the database already, there's a `record_name` column in the `continents` table that you can use.

Since we're already quite conservative about caching countries, continents, events etc. in-memory, it will be very cheap (and less redundant!) to use `Continent.c_find(continent_id).record_name` instead.